File: src/day03.hs
module Day03 (solvePartOne, solvePartTwo) where

import Data.List (elemIndex, intersect)

inputFile :: IO String
inputFile = readFile "input/day03.txt"

-- |
-- Solve part one.
--
-- >>> solvePartOne
-- 7817
solvePartOne :: IO Int
solvePartOne = solvePartOne' <$> inputFile

-- |
-- Pure solution for part one.
--
-- >>> solvePartOne' "vJrwpWtwJgWrhcsFMMfFFhFp\njqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL\nPmmdzqPrVvPwwTWBwg\nwMqvLMZHhHMvwLHjbvcjnnSBnvTQFn\nttgJtRGJQctTZtZT\nCrZsJsPPZsGzwwsLwLmpwMDw"
-- 157
solvePartOne' :: String -> Int
solvePartOne' xs = sum $ map rucksackScore $ lines xs

-- |
-- Solve part two.
--
-- >>> solvePartTwo
-- 2444
solvePartTwo :: IO Int
solvePartTwo = solvePartTwo' <$> inputFile

-- |
-- Pure solution for part two.
--
-- >>> solvePartTwo' "vJrwpWtwJgWrhcsFMMfFFhFp\njqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL\nPmmdzqPrVvPwwTWBwg\nwMqvLMZHhHMvwLHjbvcjnnSBnvTQFn\nttgJtRGJQctTZtZT\nCrZsJsPPZsGzwwsLwLmpwMDw"
-- 70
solvePartTwo' :: String -> Int
solvePartTwo' i = sum $ group'sBadges $ lines i

halve :: [a] -> ([a], [a])
halve xs = splitAt s xs
  where
    s = length xs `div` 2

rucksackScore :: String -> Int
rucksackScore xs = itemScore $ head $ uncurry intersect $ halve xs

score :: Maybe Int -> Int
score (Just s) = s + 1
score Nothing = error "Unknown item"

itemScore :: Char -> Int
itemScore c = score $ elemIndex c $ ['a' .. 'z'] ++ ['A' .. 'Z']


group'sBadges :: [String] -> [Int]
group'sBadges (x1:x2:x3:xs) = itemScore (head $ intersect x3 $ intersect x1 x2) : group'sBadges xs
group'sBadges _ = []

File: fp-ts/tests/day03.ts
import { pipe } from "fp-ts/lib/function";
import * as TE from "fp-ts/lib/TaskEither";
import * as E from "fp-ts/lib/Either";
import test from "ava";

import {
  solveSolvePartOne,
  solvePartOne,
  solvePartTwo,
  solveSolvePartTwo,
} from "../src/day03";

const exampleInput =
  "vJrwpWtwJgWrhcsFMMfFFhFp\njqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL\nPmmdzqPrVvPwwTWBwg\nwMqvLMZHhHMvwLHjbvcjnnSBnvTQFn\nttgJtRGJQctTZtZT\nCrZsJsPPZsGzwwsLwLmpwMDw";

test("[day03] test example for part one", (t) =>
  t.deepEqual(solvePartOne(exampleInput), E.right(157)));

test("[day03] solve part one", async (t) => {
  await pipe(
    solveSolvePartOne(),
    TE.match(
      (err) => t.fail(err),
      (result) => t.is(result, 7817)
    )
  )();
});

test("[day03] test example for part two", (t) =>
  t.deepEqual(solvePartTwo(exampleInput), E.right(70)));

test("[day03] solve part two", async (t) => {
  await pipe(
    solveSolvePartTwo(),
    TE.match(
      (err) => t.fail(err),
      (result) => t.is(result, 2444)
    )
  )();
});
